src/helpers/formulaHelper.js

```diff
@@ -0,0 +1,187 @@
+            while (i < tokens.length && tokens[i] >= '0' && tokens[i] <= '9') {
+                number += tokens[i];
+                i++;
+            }
+            resultStack.push({ token: parseInt(number), type: TOKEN_TYPE.Number });
+            i--;
+        }
+
+        else if (isValidNonNumberToken(tokens[i])) {
+            let tempToken = "";
+            while (i < tokens.length &&
+                (isValidNonNumberToken(tokens[i]) || (tokens[i] >= '0' && tokens[i] <= '9'))) 
+            {
+                tempToken += tokens[i];
+                i++;
+            }
+            if (constants.hasOwnProperty(tempToken)) {
+                resultStack.push({ token: constants[tempToken], type: TOKEN_TYPE.Number });
+            } else if (FUNCTION_TOKENS.includes(tempToken.toUpperCase())) {
+                parseStack.push(tempToken.toUpperCase());
+            } else {
+                resultStack.push({ token: tempToken, type: TOKEN_TYPE.Variable });
+            }
+
+            i--;
+        }
+
+        else if (tokens[i] == '(' || tokens[i] == ',') {
+            parseStack.push(tokens[i]);
+        }
+
+        else if (tokens[i] == ')') {
+            for (let char = parseStack.pop(); char != '('; char = parseStack.pop()) {
+                pushOperator(resultStack, char);
+            }
+            if (FUNCTION_TOKENS.includes(parseStack[parseStack.length - 1])) {
+                pushOperator(resultStack, parseStack.pop());
+            }
+        }
+
+        else {
+            while (parseStack.length > 0 &&
+                precedence(tokens[i]) <= precedence(parseStack[parseStack.length - 1])
+            ) {
+                pushOperator(resultStack, parseStack.pop());
+            }
+            parseStack.push(tokens[i]);
+        }
+    }
+
+    while (parseStack.length > 0) {
+        let operator = parseStack.pop();
+        pushOperator(resultStack, operator);
+    }
+
+    return resultStack;
+}
+
+function isValidNonNumberToken(char) {
+    return (
+        (char >= 'a' && char <= 'z') || 
+        (char >= 'A' && char <= 'Z') ||
+        (char == '_')
+    );
+}
+
+function precedence(operator) {
+    if (operator == '^') {
+        return 3;
+    } else if (operator == '*' || operator == '/' || operator == '÷') {
+        return 2;
+    } else if (operator == '+' || operator == '-') {
+        return 1;
+    }
+    return 0;
+}
+
+function pushOperator(resultStack, operator) {
+    // No need to return the array because array is passed by reference
+    if (FUNCTION_TOKENS.includes(operator)) {
+        resultStack.push({ token: operator, type: TOKEN_TYPE.Function });
+    } else {
+        resultStack.push({ token: operator, type: TOKEN_TYPE.Operator });
+    }
+}
+
+function evaluate(evalStack, vars) {
+    if (vars == null) {
+        throw `Missing variables`;
+    }
+
+    let resultStack = [];
+    for (let i = 0; i < evalStack.length; i++) {
+        console.log(resultStack);
```

Comment:
Do we need to log this?

Reply:
Also forgot to remove that after testing

Reply:
Pls remove this when you're done, I'll approve it after

Reply:
Removed